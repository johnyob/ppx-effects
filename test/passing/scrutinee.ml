exception%effect E : string

let f () = Stdlib.EffectHandlers.perform E
let g () = f

(* If the scrutinee is already of the form [f x], where [f] and [x] are both
   values, use [f] directly as the effect-raising function. *)
let _ =
  try f () with
  | [%effect? E, _] -> ""
;;

(* Negative test cases: *)

(* - [f] or [x] is not a value *)
let _ =
  try (Fun.id f) () with
  | [%effect? E, _] -> ""
;;

let _ =
  try f (Fun.id ()) with
  | [%effect? E, _] -> ""
;;

(* - the scrutinee isn't of the form [f x]: *)
let _ =
  try g () () with
  | [%effect? E, _] -> ""
;;

let _ =
  try
    let () = () in
    f ()
  with
  | [%effect? E, _] -> ""
;;
