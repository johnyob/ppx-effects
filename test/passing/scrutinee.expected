open! Ppx_effects_runtime.Import
type _ Ppx_effects_runtime.t +=  
  | E: string Ppx_effects_runtime.t 
let f () = perform E
let g () = f
let _ =
  Ppx_effects_runtime.Deep.try_with f ()
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | E -> Some ((fun _ -> "")) | _ -> None : continue_input
                                                                   Ppx_effects_runtime.t
                                                                   ->
                                                                   ((continue_input,
                                                                    _)
                                                                    Ppx_effects_runtime.Deep.continuation
                                                                    -> 
                                                                    _) option) in
         effc)
    }
let _ =
  Ppx_effects_runtime.Deep.try_with (fun `Unit -> (Fun.id f) ()) `Unit
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | E -> Some ((fun _ -> "")) | _ -> None : continue_input
                                                                   Ppx_effects_runtime.t
                                                                   ->
                                                                   ((continue_input,
                                                                    _)
                                                                    Ppx_effects_runtime.Deep.continuation
                                                                    -> 
                                                                    _) option) in
         effc)
    }
let _ =
  Ppx_effects_runtime.Deep.try_with (fun `Unit -> f (Fun.id ())) `Unit
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | E -> Some ((fun _ -> "")) | _ -> None : continue_input
                                                                   Ppx_effects_runtime.t
                                                                   ->
                                                                   ((continue_input,
                                                                    _)
                                                                    Ppx_effects_runtime.Deep.continuation
                                                                    -> 
                                                                    _) option) in
         effc)
    }
let _ =
  Ppx_effects_runtime.Deep.try_with (fun `Unit -> g () ()) `Unit
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | E -> Some ((fun _ -> "")) | _ -> None : continue_input
                                                                   Ppx_effects_runtime.t
                                                                   ->
                                                                   ((continue_input,
                                                                    _)
                                                                    Ppx_effects_runtime.Deep.continuation
                                                                    -> 
                                                                    _) option) in
         effc)
    }
let _ =
  Ppx_effects_runtime.Deep.try_with (fun `Unit -> let () = () in f ()) `Unit
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function | E -> Some ((fun _ -> "")) | _ -> None : continue_input
                                                                   Ppx_effects_runtime.t
                                                                   ->
                                                                   ((continue_input,
                                                                    _)
                                                                    Ppx_effects_runtime.Deep.continuation
                                                                    -> 
                                                                    _) option) in
         effc)
    }
